src/lexer.mll

```diff
@@ -271,8 +271,61 @@ let raw_markup_target =
 let language_tag_char =
   ['a'-'z' 'A'-'Z' '0'-'9' '_' '-' ]
 
+rule reference_paren_content input start depth buffer = parse 
+  | '('
+    {
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+       reference_paren_content input start (depth + 1) buffer lexbuf }
+  | ')'
+    { 
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+      if depth = 0 then
+        ()
+      else
+        ( reference_paren_content input start (depth  - 1) buffer lexbuf ) }
+  | eof 
+    { warning
+        input
+        ~start_offset:(Lexing.lexeme_end lexbuf)
+        (Parse_error.not_allowed
+          ~what:(Token.describe `End)
+          ~in_what:(Token.describe (reference_token start ""))) 
+    }
+  | _
+    { 
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+      reference_paren_content input start depth buffer lexbuf }
 
-rule token input = parse
+and reference_content input start buffer = parse
+  | '}' 
+    {
+      Buffer.contents buffer
+    }
+  | '('
+    {
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+      ((reference_paren_content input start 0 buffer lexbuf)) ;
+      reference_content input start buffer lexbuf
+    }
+  | '"' [^ '"']* '"'
+    {
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+      reference_content input start buffer lexbuf
+    }
+  | eof 
+    { warning
+        input
+        ~start_offset:(Lexing.lexeme_end lexbuf)
+        (Parse_error.not_allowed
+          ~what:(Token.describe `End)
+          ~in_what:(Token.describe (reference_token start "")));
+        Buffer.contents buffer }
+  | _
```

Comment:
We could have a rule for `)` to warn about unbalanced parentheses.

Reply:
What do you think about this ? For example `{!)}` reports no warning.

Reply:
Also, maybe unrelated to this PR, it might be time to define what characters are allowed in references outside of `""` or `()`. For example `{` isn't at the moment.